### PR TITLE
Adding UserReportDoc to database model. 

### PR DESCRIPTION
**Pull request by JCantu248**
Adding UserReportDocs to the PCA database.

## 🗣 Description ##

A document has been added to the PCA database to track when customers reply to campaign emails.

A new command `pca-import --user-report`  was added to allow UserReportDocs to be brought into the PCA database.

A new command `pca-output --output-user-report`  was added to allow a user to output UserReportDocs for a given customer and assessment. 

## 💭 Motivation and context ##

Kelly Thiele and the PCA team want to be able to track when a customer responds to a phishing email generated by the PCA team. They want to store records of when the customer first responded to their campaign, and how many times they have done so. This required changes to the database model which have been made here.
## 🧪 Testing ##

These changes were tested by using the `pca-import --user-report` and `pca-tool --output-user-report ` commands. 

I searched the development PCA database to find corresponding customer ids, assessment ids, and campaign ids. For security purposes the exact data cannot be shown. I used that data to build a JSON object of the following design: 

```
{
	"customer":  ,
	"assessment": ,
	"campaign": 
	"first_report": ,
	"total_num_reports": 

} 
```

I passed the JSON file to the importer command, and then used the `pca-tool --output-user-report` command to view the loaded data. This demonstrates the correct operation of reading and writing to the database using the new document. 

I tested the importer with malformed JSON input, and tested it against previously loaded data to ensure it would block the attempt without the --force flag. 

The output command was tested to ensure correct data was displayed, according to the provided input. 

Made some changes to the code and retested to make sure that when forcing records into the database, it would rewrite over the previously entered data. 

## ✅ Checklist ##

* [x] This PR has an informative and human-readable title.
* [x] Changes are limited to a single goal - _eschew scope creep!_
* [x] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
* [x] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
* [x] All relevant repo and/or project documentation has been updated
      to reflect the changes in this PR.
* [x] All new and existing tests pass.

### CODE REVIEW

File: pca/db/database_model.py [hunk 1]
```diff
@@ -84,7 +86,7 @@ class Meta:
         collection_name = CUSTOMER_COLLECTION
         final = True  # so we don't get a '_cls' field in these documents (e.g. "_cls" : "__main__.Customer")
 
-    def get_all_customers():
+    def get_all_customers(self):
```
Comment (jsf9k) [inline pca/db/database_model.py]: Is this change necessary?  Can this method remain static?

**Review by jsf9k — COMMENTED** [review]
I have one minor question.

File: pca/db/database_model.py (hunk 1 above)
Comment (JCantu248) [reply pca/db/database_model.py]: Yes, that can remain static. I'll make the change.

**Comment by dav3r** [comment]
@JCantu248 - Can you please fill out the "Testing" section in the PR description so that we can understand how you tested these changes?

Also, please take care of all of the items in the "Checklist" section.  Ideally, those should all be checked before requesting a PR review.  If any of them do not apply to this particular PR, you can delete them (just don't delete items from that list because you don't feel like doing them).

File: pca/db/database_model.py [hunk 2]
```diff
@@ -417,3 +419,36 @@ def save(self, *args, **kwargs):
...
         super(ApplicationDoc, self).save(*args, **kwargs)
+
+class UserReportDoc(RootDoc):
+    """Provides schema for User Report doc."""
+
+    customer = fields.ReferenceField(CustomerDoc, required=True)
+    assessment = fields.ReferenceField(AssessmentDoc, required=True)
+    campaign = fields.ReferenceField(CampaignDoc, required=True)
+    first_report = fields.DateTimeField(blank=True)
+    total_num_reports = fields.IntegerField(blank=True)
```
Comment (dav3r) [inline pca/db/database_model.py]: When would `total_num_reports` ever be blank?  When one of these documents is initially created, shouldn't that field be initialized to either zero or one (depending on when you plan to create these docs)?

File: bin/pca-import [hunk 3]
```diff
@@ -773,6 +777,48 @@ def import_template_file(template_filename, template_id=None, force=False):
...
+        print("Cannot import, campaign_id does not exist: {}".format(
+            request["campaign"]
+            )
+        )
+        return False
+    if not force and user_report_already_exists(request["customer"],
+                                                request["assessment"],
+                                                request["campaign"]):
+        print("Cannot import, user_report already exists: {} {} {}".format(
+            request["customer"], request["assessment"], request["campaign"]))
```
Comment (dav3r) [inline bin/pca-import]: This error message text should probably make it clearer to the end user that the values being output correspond to `customer_id`, `assessment_id`, and `campaign_id`.

File: bin/pca-import [hunk 4]
```diff
@@ -170,27 +172,29 @@ def import_assessment(request, source, force=False):
...
         return import_customer(request, filename, force)
-    elif assessmentImport:
+    elif assessment_import:
         return import_assessment(request, filename, force)
+    elif user_report_import:
+        return import_user_report(request, filename, force)
 
 
-def import_json_stdin(force, customerImport, assessmentImport):
+def import_json_stdin(force, customer_import, assessment_import):
```
Comment (dav3r) [inline bin/pca-import]: This function also needs to be updated to handle `user_report_import`.  You should also test that this works as intended by passing in the user report JSON via `stdin`.

File: bin/pca-tool [hunk 5]
```diff
@@ -419,6 +420,21 @@ def delete_image(image_id):
     return True
 
 
+def output_user_report(customer_id, assessment_id):
```
Comment (dav3r) [inline bin/pca-tool]: Since this function can output more than one user report, I'd call it `output_user_reports`:
```suggestion
def output_user_reports(customer_id, assessment_id):
```

Don't forget to update any other references to this function accordingly.

File: pca/db/database_model.py [hunk 6]
```diff
@@ -417,3 +419,22 @@ def save(self, *args, **kwargs):
         else:
             self.internal_ip_int = None  # IPv6 integer representations are too large to store in MongoDB currently
         super(ApplicationDoc, self).save(*args, **kwargs)
+
+class UserReportDoc(RootDoc):
+    """Provides schema for User Report doc."""
```
Comment (dav3r) [inline pca/db/database_model.py]: I'd prefer to see a bit more explanation in here about what this collection and the `first_report` date actually represent, since it's not clear from the code.

File: bin/pca-tool [hunk 7]
```diff
@@ -419,6 +420,21 @@ def delete_image(image_id):
...
+    user_reports = list(UserReportDoc.objects.raw({"customer": customer_id,
+                                                   "assessment": assessment_id}).order_by([("_id", pymongo.ASCENDING)]))
+    if user_reports:
+        print("customer_id", "campaign_id", "assessment_id", "first_report", "total_num_reports")
+        for user_report in user_reports:
+            print('"{0.customer._id}", "{0.assessment._id}", "{0.campaign._id}", {0.first_report}, {0.total_num_reports}'.format(
+                user_report
+            ))
```
Comment (dav3r) [inline bin/pca-tool]: The order of your field headings does not match the order of the data you are printing out (campaign and assessment IDs are reversed).

File: bin/pca-tool [hunk 8]
```diff
@@ -419,6 +420,21 @@ def delete_image(image_id):
...
+        return True
+    else:
+        print("No User Reports for customer {} and assessment {}".format(customer_id, assessment_id))
```
Comment (dav3r) [inline bin/pca-tool]: No need to capitalize here:
```suggestion
        print("No user reports for customer {} and assessment {}".format(customer_id, assessment_id))
```

**Review by dav3r — CHANGES_REQUESTED** [review]
@JCantu248 - This is looking good, but please take a look at my comments and the changes I requested.

Also, please update the description of this PR to mention that you also added new options to both `pca-import` and `pca-tool`.  Even though you mentioned that in the testing, it's helpful to have the PR description contain a concise overview of the PR changes.

**Comment by dav3r** [comment]
One other thing @JCantu248 - please assign yourself and add the `improvement` label to this PR.  Thanks!

File: pca/db/database_model.py (hunk 2 above)
Comment (JCantu248) [reply pca/db/database_model.py]: You are correct. Its been changed to blank=False

File: bin/pca-import (hunk 4 above)
Comment (JCantu248) [reply bin/pca-import]: Tested with stdin. It is capable of adding a document to the database by console input

**Comment by dav3r** [comment]
@JCantu248 - In the future, please let the person who initiated each conversation be the one who marks the conversation as "Resolved".  That is how we typically know which items have not yet been resolved to our satisfaction.

Also, it is very helpful if you reply to each conversation with the commit URL where the issue was corrected (e.g. https://github.com/cisagov/pca-core/pull/5/commits/3733c13054a3c9a1a54dbb46983c9070043c733e).  Related to that, we prefer multiple, smaller commits, where each commit addresses a particular issue, rather than a single large commit that covers a wide variety of issues.  These two items combined make it much easier for a reviewer to re-review any updates that you made.

File: bin/pca-import [hunk 9]
```diff
@@ -773,6 +779,59 @@ def import_template_file(template_filename, template_id=None, force=False):
...
+
+    user_report = UserReportDoc.find_by_customer_assessment_campaign(request["customer"],
+                                                                     request["assessment"],
+                                                                     request["campaign"])
+    if user_report:
+        if force:
+            user_report_id = user_report._id
+        else:
+            print(
+                "Cannot import, user_report already exist - Customer: {} Assessment: {} Campaign: {}".format(
```
Comment (dav3r) [inline bin/pca-import]: Fix grammar and match other error messages:
```suggestion
                "Cannot import, user_report already exists - customer_id: {} assessment_id: {} campaign_id: {}".format(
```

File: bin/pca-import (hunk 4 above)
Comment (dav3r) [reply bin/pca-import]: Looks good, thanks!

File: pca/db/database_model.py [hunk 10]
```diff
@@ -417,3 +419,38 @@ def save(self, *args, **kwargs):
...
             self.internal_ip_int = None  # IPv6 integer representations are too large to store in MongoDB currently
         super(ApplicationDoc, self).save(*args, **kwargs)
+
+class UserReportDoc(RootDoc):
+    """Provides schema for User Report doc.
+    customer - a customer ID that corresponds to a CustomerDoc
+    assessment - an assessment ID that corresponds to an AssessmentDoc
+    campaign - a campaign ID that corresponds to a CampaignDoc
+    first_report - the first time a user responds to a phishing email. Optional, may not be available from GoPhish
+    total_num_reports - the total number of responses (reports) for a campaign.
```
Comment (dav3r) [inline pca/db/database_model.py]: How about this?  I don't think you need the "Optional..." bit, since that is covered by the field definition below.
```suggestion
    first_report - the timestamp when the first user click ("report") was received during a phishing campaign
    total_num_reports - the total number of user clicks ("reports") for a campaign
```

File: bin/pca-import (hunk 9 above)
Comment (JCantu248) [reply bin/pca-import]: Suggestion was added in 86972531a8facb739a161eecba3a8c1e546aca54

File: pca/db/database_model.py (hunk 10 above)
Comment (JCantu248) [reply pca/db/database_model.py]: Changes applied in 76ffea3eb11ba5c77d2593759504e71da3353c4e

File: pca/db/database_model.py (hunk 6 above)
Comment (dav3r) [reply pca/db/database_model.py]: Looking good now after https://github.com/cisagov/pca-core/pull/5/commits/3733c13054a3c9a1a54dbb46983c9070043c733e and https://github.com/cisagov/pca-core/pull/5/commits/76ffea3eb11ba5c77d2593759504e71da3353c4e

**Review by dav3r — APPROVED** [review]
I think this PR is good to go now - thanks for making the requested changes.  👍 

The one last thing that you should do is to delete the checklist item (in the PR description) that doesn't apply:  "Tests have been added and/or modified to cover the changes in this PR."

File: bin/pca-import [hunk 11]
```diff
@@ -170,27 +172,31 @@ def import_assessment(request, source, force=False):
     return True
 
 
-def import_json_file(filename, force, customerImport, assessmentImport):
+def import_json_file(filename, force, customer_import, assessment_import, user_report_import):
```
Comment (jsf9k) [inline bin/pca-import]: Do you intend this function to perform only one type of import?  What if someone passes `customer_import = True` and `assessment_import = True`?  I see that the code will only perform the customer import, but this may not be what the user expects.

The best way to address this would be for the function to accept a single `enum` value (`enums` are backported to Python 2.7 via [this `pip` package](https://pypi.org/project/enum34/)) value instead of the three Boolean values.  That said, I'm willing to approve if you create a docstring for this function describing that it expects _exactly one_ of the three Booleans to be `True`.

File: bin/pca-import [hunk 12]
```diff
@@ -170,27 +172,31 @@ def import_assessment(request, source, force=False):
...
         return import_customer(request, filename, force)
-    elif assessmentImport:
+    elif assessment_import:
         return import_assessment(request, filename, force)
+    elif user_report_import:
+        return import_user_report(request, filename, force)
 
 
-def import_json_stdin(force, customerImport, assessmentImport):
+def import_json_stdin(force, customer_import, assessment_import, user_report_import):
```
Comment (jsf9k) [inline bin/pca-import]: The comment above about `import_json_file()` also applies here.

File: pca/db/database_model.py [hunk 13]
```diff
@@ -417,3 +419,38 @@ def save(self, *args, **kwargs):
...
+    campaign = fields.ReferenceField(CampaignDoc, required=True)
+    first_report = fields.DateTimeField(blank=True)
+    total_num_reports = fields.IntegerField(blank=False)
+
+    class Meta:
+        """Meta class for UserReportDoc."""
+
+        collection_name = USER_REPORT_COLLECTION
+        final = True
+        ignore_unknown_fields = True  # TODO see if this can be inherited from RootDoc
```
Comment (jsf9k) [inline pca/db/database_model.py]: You should create an issue for this TODO, and include a link to the issue in this comment.

**Review by jsf9k — CHANGES_REQUESTED** [review]
Strong work @JCantu248!

I made a few minor requests.  Please re-request my review after you address them.

File: bin/pca-import (hunk 11 above)
Comment (JCantu248) [reply bin/pca-import]: See a8014ba675ba14a104152964cb72cf7ce0bb7127 for my changes. I made it an enum value instead. It cleaned up the code somewhat

File: bin/pca-import (hunk 12 above)
Comment (JCantu248) [reply bin/pca-import]: See a8014ba675ba14a104152964cb72cf7ce0bb7127

File: pca/db/database_model.py (hunk 13 above)
Comment (JCantu248) [reply pca/db/database_model.py]: Issue can be found here: [Issue](https://github.com/cisagov/pca-core/issues/7)

Comment (jsf9k) [reply pca/db/database_model.py]: @JCantu248, please reference #7 in this comment.  Then re-request my review and I'll approve.

**Review by jsf9k — CHANGES_REQUESTED** [review]
There is still one outstanding issue that should be resolved, then I can approve this PR.

File: pca/db/database_model.py [hunk 14]
```diff
@@ -417,3 +419,38 @@ def save(self, *args, **kwargs):
...
+    assessment - an assessment ID that corresponds to an AssessmentDoc
+    campaign - a campaign ID that corresponds to a CampaignDoc
+    first_report - the timestamp when the first user click ("report") was received during a phishing campaign
+    total_num_reports - the total number of user clicks ("reports") for a campaign
+    """
+    customer = fields.ReferenceField(CustomerDoc, required=True)
+    assessment = fields.ReferenceField(AssessmentDoc, required=True)
+    campaign = fields.ReferenceField(CampaignDoc, required=True)
+    first_report = fields.DateTimeField(blank=True)
+    total_num_reports = fields.IntegerField(blank=False)
```
Comment (dav3r) [inline pca/db/database_model.py]: @JCantu248 - Sorry to go back to this, but is there any reason why this can't be:
```suggestion
    total_num_reports = fields.IntegerField(required=True)
```

Comment (JCantu248) [reply pca/db/database_model.py]: I set it that way because I wanted to account for having no reports generated, but then I have first_report set to true, so they are contradictory. I may want to make first_report false as well in the event that no user responds to a phishing email.

File: pca/db/database_model.py (hunk 13 above)
Comment (JCantu248) [reply pca/db/database_model.py:442]: Umm okay.  #7

Comment (JCantu248) [reply pca/db/database_model.py]: See commit abaedb5bdfa316bbf3ba6c0d93c873b56b431977 for code changes

**Review by jsf9k — APPROVED** [review]
Thanks for taking care of all those changes @JCantu248!
